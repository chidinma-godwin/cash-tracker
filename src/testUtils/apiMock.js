import {
  loginEndpoint,
  signupEndpoint,
  transactionEndpoint,
  userEndpoint,
} from 'constants/endpoints';
import { createServer, Model } from 'miragejs';

export default function makeServer(environment = 'test') {
  return createServer({
    environment,
    models: {
      user: Model,
      transaction: Model,
    },
    routes() {
      this.get(userEndpoint, () => ({
        user: {
          id: '1',
          username: 'Authenticated User',
          email: 'user@email.com',
          clientsEmail: ['client1@email.com'],
          pendingInvitations: ['invite@email.com'],
          pendingRequests: ['pending1@email.com', 'pending2@email.com'],
        },
      }));
      this.post(signupEndpoint, (schema, request) => {
        const attrs = JSON.parse(request.requestBody);
        return schema.users.create(attrs);
      });
      this.post(loginEndpoint, (schema, request) => {
        const { email, password } = JSON.parse(request.requestBody);
        return schema.users.findBy({ email, password });
      });
      this.get(`${transactionEndpoint}/:id`, (schema, request) => {
        const { id } = request.params;
        return schema.transactions.where(transaction =>
          transaction.participants.includes(id)
        );
      });
      this.post(transactionEndpoint, (schema, request) => {
        const attrs = JSON.parse(request.requestBody);
        return schema.transactions.create(attrs);
      });
    },
  });
}
