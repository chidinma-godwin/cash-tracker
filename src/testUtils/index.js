import React from 'react';
import { SWRConfig } from 'swr';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';

// eslint-disable-next-line react/prop-types
const Wrapper = ({ children }) => (
  <SWRConfig value={{ dedupingInterval: 0 }}>{children}</SWRConfig>
);

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options });

const customRenderHook = hook => renderHook(hook, { wrapper: Wrapper });

export * from '@testing-library/react';
export { customRenderHook as renderHook };
export { customRender as render };
