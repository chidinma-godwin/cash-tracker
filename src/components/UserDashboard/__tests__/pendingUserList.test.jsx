import userEvent from '@testing-library/user-event';

// CashTracker imports
import { render, screen } from 'testUtils';
import PendingUserList from '../pendingUserList';

describe('PendingUserList component', () => {
  const requiredProps = {
    pendingInvitations: ['test@email.com', 'another@email.com'],
    handleTransactionAccount: jest.fn(),
  };
  it('matches snapshot', () => {
    const { asFragment } = render(<PendingUserList {...requiredProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('calls handleTransactionAccount props with the right data when add button is clicked', () => {
    render(<PendingUserList {...requiredProps} />);

    const addButton = screen.getAllByRole('button', {
      name: /confirm account/i,
    });
    userEvent.click(addButton[1]);

    expect(requiredProps.handleTransactionAccount).toHaveBeenCalledWith(
      'another@email.com',
      'add'
    );
  });
  it('calls handleTransactionAccount props with the right data when delete button is clicked', () => {
    render(<PendingUserList {...requiredProps} />);

    const deleteButton = screen.getAllByRole('button', {
      name: /delete request/i,
    });
    userEvent.click(deleteButton[0]);

    expect(requiredProps.handleTransactionAccount).toHaveBeenCalledWith(
      'test@email.com',
      'delete'
    );
  });
});
