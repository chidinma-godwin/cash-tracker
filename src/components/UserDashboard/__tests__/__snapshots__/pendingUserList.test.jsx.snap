// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingUserList component matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation8 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
          >
            Pending Invitations
          </span>
          <span
            class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            Accept the request to start transactions, otherwise you can decline the request
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <li
          class="MuiListItem-container"
        >
          <div
            class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          >
            <div
              class="MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                test@email.com
              </span>
            </div>
          </div>
          <div
            class="MuiListItemSecondaryAction-root"
          >
            <button
              aria-label="confirm account"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="delete request"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
        <li
          class="MuiListItem-container"
        >
          <div
            class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          >
            <div
              class="MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                another@email.com
              </span>
            </div>
          </div>
          <div
            class="MuiListItemSecondaryAction-root"
          >
            <button
              aria-label="confirm account"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="delete request"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
      </div>
    </div>
  </ul>
</DocumentFragment>
`;
